### PR TITLE
Feature/#131TurnDotNet8

### PR DESCRIPTION
**Pull request by Ozodbek-Anvarjonov**

### CODE REVIEW

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Validators/EmailMessageValidator.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+namespace TruckWorld.Infrastructure.Common.Validators;
+
+public class EmailMessageValidator : AbstractValidator<EmailMessage>
+{
+    public EmailMessageValidator()
+    {
+        RuleSet(NotificationEvent.OnRedering.ToString(),
+            () =>
+            {
+                RuleFor(message => message.Variables).NotNull();
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Validators/EmailMessageValidator.cs]: So the first check must be template right ?

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/SmsSenderService.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+using TruckWorld.Domain.Extensions;
+
+namespace TruckWorld.Infrastructure.Common.Notifications.Services;
+
+public class SmsSenderService : ISmsSenderService
+{
+    private readonly IEnumerable<ISmsSenderBroker> _smsSenderBroker;
+    private readonly IValidator<SmsMessage> _smsSenderValidator;
+
+    public SmsSenderService(IEnumerable<ISmsSenderBroker> smsSenderBroker, IValidator<SmsMessage> smsSenderValidator)
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/SmsSenderService.cs:15]: Use primary constructor

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Validators/SmsMessageValidator.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            {
+                RuleFor(message => message.Template).NotNull();
+                RuleFor(message => message.Variables).NotNull();
+            });
+
+        RuleSet(NotificationEvent.OnSending.ToString(),
+            () =>
+            {
+                RuleFor(message => message.SenderPhoneNumber).NotNull().NotEmpty();
+                RuleFor(message => message.RecieverPhoneNumber).NotNull().NotEmpty();
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Validators/SmsMessageValidator.cs]: Check with regex from external validation ( inject `ValidationSettings` into validator )

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Validators/EmailMessageValidator.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                RuleFor(message => message.Variables).NotNull();
+            });
+
+        RuleSet(NotificationEvent.OnSending.ToString(),
+            () =>
+            {
+                RuleFor(message => message.RecieverEmailAddress).NotNull().NotEmpty();
+                RuleFor(message => message.SenderEmailAddress).NotNull().NotEmpty();
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Validators/EmailMessageValidator.cs]: Check with regex from external validation ( inject `ValidationSettings` into validator )

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+using TruckWorld.Domain.Extensions;
+
+namespace TruckWorld.Infrastructure.Common.Notifications.Services;
+
+public class EmailSenderService : IEmailSenderService
+{
+    private readonly IEnumerable<IEmailSenderBroker> _mailSenderBroker;
+    private readonly IValidator<EmailMessage> _emailMessageValidator;
+
+    public EmailSenderService(IEnumerable<IEmailSenderBroker> emailSenderBrokers, IValidator<EmailMessage> validator)
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs:15]: Use primary constructor

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/TwilioSmsSenderBroker.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+using Twilio.Types;
+
+
+namespace TruckWorld.Infrastructure.Common.Notifications.Brokers;
+
+public class TwilioSmsSenderBroker : ISmsSenderBroker
+{
+    private readonly TwilioSmsSenderSettings _settings;
+
+    public TwilioSmsSenderBroker(IOptions<TwilioSmsSenderSettings> settings) =>
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/TwilioSmsSenderBroker.cs:16]: Use primary constructor

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/SmtpEmailSenderBroker.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    public async ValueTask<bool> SendAsync(EmailMessage emailMessage, CancellationToken cancellationToken = default)
+    {
+        emailMessage.SenderEmailAddress ??= _settings.CredentialAddress;
+
+        var mail = new MailMessage(emailMessage.SenderEmailAddress, emailMessage.RecieverEmailAddress);
+
+        mail.Subject = emailMessage.Subject;
+        mail.Body = emailMessage.Body;
+        mail.IsBodyHtml = true;
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/SmtpEmailSenderBroker.cs:25]: Actually add this to email template and set from email template

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/SmtpEmailSenderBroker.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+using TruckWorld.Infrastructure.Common.Settings;
+
+namespace TruckWorld.Infrastructure.Common.Notifications.Brokers;
+
+public class SmtpEmailSenderBroker : IEmailSenderBroker
+{
+    private readonly SmtpEmailSenderSettings _settings;
+
+    public SmtpEmailSenderBroker(IOptions<SmtpEmailSenderSettings> smtpEmailSenderSettings) =>
+        _settings = smtpEmailSenderSettings.Value;
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/SmtpEmailSenderBroker.cs:15]: Use primary constructor

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Mapper/NotificationMessageMapper.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+﻿using AutoMapper;
+
+namespace TruckWorld.Infrastructure.Common.Mapper;
+
+public class NotificationMessageMapper : Profile
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Mapper/NotificationMessageMapper.cs:5]: Is it supposed to map from notification message to sms message and email message ?

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationEvent.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+/// <summary>
+/// Represents a notification event
+/// </summary>
+public enum NotificationEvent
+{
+    /// <summary>
+    /// status in rendering the notification
+    /// </summary>
+    OnRedering,
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationEvent.cs:11]: Fix to OnRendering

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Entities/SmsTemplate.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿using TruckWorld.Domain.Common.Entities;
+using TruckWorld.Domain.Enums;
+
+namespace TruckWorld.Domain.Entities;
+
+public class SmsTemplate : IEntity
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Entities/SmsTemplate.cs:6]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Entities/EmailTemplate.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
+﻿using TruckWorld.Domain.Common.Entities;
+using TruckWorld.Domain.Enums;
+
+namespace TruckWorld.Domain.Entities;
+
+public class EmailTemplate : IEntity
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Entities/EmailTemplate.cs:6]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Services/ISmsSenderService.cs [hunk 13]
```diff
@@ -0,0 +1,8 @@
+﻿using TruckWorld.Application.Common.Models;
+
+namespace TruckWorld.Application.Common.Services;
+
+public interface ISmsSenderService
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Services/ISmsSenderService.cs:7]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Services/IEmailSenderService.cs [hunk 14]
```diff
@@ -0,0 +1,8 @@
+﻿using TruckWorld.Application.Common.Models;
+
+namespace TruckWorld.Application.Common.Services;
+
+public interface IEmailSenderService
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Services/IEmailSenderService.cs:8]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Models/SmsMessage.cs [hunk 15]
```diff
@@ -0,0 +1,14 @@
+﻿using TruckWorld.Domain.Entities;
+
+namespace TruckWorld.Application.Common.Models;
+
+public class SmsMessage : NotificationMessage
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Models/SmsMessage.cs:8]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Models/NotificationMessage.cs [hunk 16]
```diff
@@ -0,0 +1,14 @@
+﻿namespace TruckWorld.Application.Common.Models;
+
+public class NotificationMessage
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Models/NotificationMessage.cs:6]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Models/EmailMessage.cs [hunk 17]
```diff
@@ -0,0 +1,16 @@
+﻿using TruckWorld.Domain.Entities;
+
+namespace TruckWorld.Application.Common.Models;
+
+public class EmailMessage : NotificationMessage
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Models/EmailMessage.cs:8]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Brokers/ISmsSenderBroker.cs [hunk 18]
```diff
@@ -0,0 +1,8 @@
+﻿using TruckWorld.Application.Common.Models;
+
+namespace TruckWorld.Application.Common.Brokers;
+
+public interface ISmsSenderBroker
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Brokers/ISmsSenderBroker.cs:8]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Brokers/IEmailSenderBroker.cs [hunk 19]
```diff
@@ -0,0 +1,8 @@
+﻿using TruckWorld.Application.Common.Models;
+
+namespace TruckWorld.Application.Common.Brokers;
+
+public interface IEmailSenderBroker
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Brokers/IEmailSenderBroker.cs:8]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Settings/SmtpEmailSenderSettings.cs [hunk 20]
```diff
@@ -0,0 +1,12 @@
+﻿namespace TruckWorld.Infrastructure.Common.Settings;
+
+public class SmtpEmailSenderSettings
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Settings/SmtpEmailSenderSettings.cs:6]: Add docs

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Settings/TwilioSmsSenderSettings.cs [hunk 21]
```diff
@@ -0,0 +1,10 @@
+﻿namespace TruckWorld.Infrastructure.Common.Settings;
+
+public class TwilioSmsSenderSettings
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Settings/TwilioSmsSenderSettings.cs:6]: Add docs

**Review by sulton-max — CHANGES_REQUESTED** [review]

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs (hunk 5 above)
Comment (Ozodbek-Anvarjonov) [reply src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs:15]: project in .Net 7

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/TwilioSmsSenderBroker.cs [hunk 22]
```diff
@@ -0,0 +1,30 @@
+﻿using Microsoft.Extensions.Options;
+using TruckWorld.Application.Common.Brokers;
+using TruckWorld.Application.Common.Models;
+using TruckWorld.Infrastructure.Common.Settings;
+using Twilio;
+using Twilio.Rest.Api.V2010.Account;
+using Twilio.Types;
+
+
+namespace TruckWorld.Infrastructure.Common.Notifications.Brokers;
```
Comment (botirjonabdumannonovv) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/TwilioSmsSenderBroker.cs:10]: Remove unnecessary row

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+using TruckWorld.Domain.Enums;
+using TruckWorld.Domain.Extensions;
+
+namespace TruckWorld.Infrastructure.Common.Notifications.Services;
+
+public class EmailSenderService : IEmailSenderService
+{
+    private readonly IEnumerable<IEmailSenderBroker> _mailSenderBroker;
+    private readonly IValidator<EmailMessage> _emailMessageValidator;
+
```
Comment (botirjonabdumannonovv) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs:14]: Use .NET 8 Primary constructor

**Pull request closed without merging**